### PR TITLE
update .d.ts files

### PR DESCRIPTION
**Pull request by moheng233**
update .d.ts files

### CODE REVIEW

**Comment by ItalyPaleAle** [comment]
Thank you for starting this, it's a great starting point! There are a few things that needs to be changed however. I'm not a TypeScript expert so I'm not able to do it myself sadly 😞

File: Router.d.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export interface RouteDetail {
+    route: string | RegExp;
+    location: string;
+    querystring: string;
+    userData?: {};
+    component?: {};
```
Comment (ItalyPaleAle) [inline Router.d.ts:6]: `component` is a `SvelteComponent`

File: Router.d.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    name?: string;
+}
```
Comment (ItalyPaleAle) [inline Router.d.ts:8]: Sadly, the problem is that when you add a `.d.ts` file, you need to document the types for all exports, or editors will miss the JSDoc comments.

In this file, there are some definitions missing:

- `push`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L94
- `pop`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L112
- `replace`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L125
- `link`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L158

Also, I don't know if types need to be exported for Svelte stores too? This component is exporting some Svelte stores:

- `loc`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L55
- `location`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L75
- `querystring`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L83

What about props? Do those need to be in the typing file too?

- `routes`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L236
- `prefix`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L241
- `restoreScrollState`: https://github.com/ItalyPaleAle/svelte-spa-router/blob/master/Router.svelte#L247

File: wrap.d.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import {SvelteComponent} from 'svelte'
+import {RouteDetail} from './Router'
+
+interface WrappedComponent {
+    component: SvelteComponent
+    conditions?: RoutePrecondition[]
+    props?: {}
+    userData?: {}
+    _sveltesparouter: boolean
```
Comment (ItalyPaleAle) [inline wrap.d.ts:9]: Remove `_sveltesparouter` as it's an internal property only

File: wrap.d.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+type RoutePrecondition = (detail: any) => boolean
```
Comment (ItalyPaleAle) [inline wrap.d.ts:12]: Actually this can be an async function too, so it can either return `boolean` or `Promise<boolean>`

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: wrap.d.ts (hunk 4 above)
Comment (moheng233) [reply wrap.d.ts:12]: Okay, I will make changes to it.
After all, the absence of .d.ts is a nightmare for people who like to develop with ts.

File: Router.d.ts (hunk 2 above)
Comment (moheng233) [reply Router.d.ts:8]: I don't really know,
So router. D.ts will be generated by svelte2tsx

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by ItalyPaleAle** [comment]
Ok I made the changes and I think this works (tested it in another repo). I'll squash then merge!

**Comment by BogdanDarius** [comment]
Add some examples with typescript maybe?

**Comment by ItalyPaleAle** [comment]
I'm actually working on some start templates for svelte-spa-router, so you can start a new project with a GitHub template repo! But I'm blocked by this: https://github.com/sveltejs/template/issues/78

**Comment by gdamjan** [comment]
with 3.0.4 I'm getting this error:

```
Getting Svelte diagnostics...
====================================

/home/damjan/my-stuff/irclogs/svelte/src/App.svelte:46:3
Error: JSX element class does not support attributes because it does not have a '$$prop_def' property. (ts)
<main>
  <Router {routes} />
</main>


/home/damjan/my-stuff/irclogs/svelte/src/App.svelte:46:4
Error: Type definitions are missing for this Svelte Component. It needs a class definition with at least the property '$$prop_def' which should contain a map of input property definitions.
Example:
class ComponentName { $$prop_def: { propertyName: string; } }

'Router' cannot be used as a JSX component.
  Its instance type 'Router' is not a valid JSX element.
    Property '$$prop_def' is missing in type 'Router' but required in type 'ElementClass'. (ts)
<main>
  <Router {routes} />
</main>

```

**Comment by ItalyPaleAle** [comment]
@gdamjan thanks, I'm going to look into this issue this afternoon

**Comment by ItalyPaleAle** [comment]
Turns out I had completely f---d up this time and released something that I should have tested MUCH more thoroughly. I'm working on an actual fix, this time testing it properly, and will be released tonight.

I'm sorry for this, I should have known better.

**Comment by ItalyPaleAle** [comment]
Ok v3.0.5 is out. These typings should now be correct. Apologies again for the mess

**Comment by gdamjan** [comment]
excellent, thanks a lot

**Pull request merged**